tabledesc: add validation for MutationJobs

postamar opened the pull request:
Previously, the contents of this slice were almost not validated. This has become a problem because the declarative schema changer relies on it to forbid schema changes concurrent with the legacy schema changer. This patch addresses this by adding the missing validation logic. Henceforth a nonempty MutationJobs slice implies a nonempty Mutations slice.

Fixes #87945.

Release note: None

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=91747) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=91747) before we can accept your contribution.<br/><hr/>**Marius Posta** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=91747) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91747)
<!-- Reviewable:end -->

Commit:
tabledesc: add validation for MutationJobs

Previously, the contents of this slice were almost not validated. This
has become a problem because the declarative schema changer relies on it
to forbid schema changes concurrent with the legacy schema changer. This
patch addresses this by adding the missing validation logic. Henceforth
a nonempty MutationJobs slice implies a nonempty Mutations slice.

Fixes #87945.

Release note: None

Xiang-Gu commented on pkg/sql/catalog/tabledesc/validate.go:
```diff
@@ -675,13 +675,31 @@ func (desc *wrapper) ValidateSelf(vea catalog.ValidationErrorAccumulator) {
+				continue
+			}
+			if jobID, found := jobIDs[mj.MutationID]; found {
+				vea.Report(errors.AssertionFailedf("duplicate job mappings %d and %d for mutation ID %d",
+					jobID, mj.JobID, mj.MutationID))
```
nit: can the error message be easier to understand, something like "two jobs exists for the same mutation ID"? 

TIL: one mutation ID corresponds to one or more mutations on the mutation slice (i.e. they all share the same mutation ID as a result of the same DDL statement), but one mutation ID always corresponds to one job ID.

Xiang-Gu commented on pkg/sql/catalog/tabledesc/validate.go:
```diff
@@ -675,13 +675,31 @@ func (desc *wrapper) ValidateSelf(vea catalog.ValidationErrorAccumulator) {
 				vea.Report(err)
 				mutationsHaveErrs = true
 			}
 		}
```
nit: Can we add a comment here that says something like "validate one mutation ID corresponds to exactly one mutation job"?

Xiang-Gu commented on pkg/sql/truncate.go:
```diff
@@ -215,9 +215,15 @@ func (p *planner) truncateTable(ctx context.Context, id descpb.ID, jobDesc strin
-	version := p.ExecCfg().Settings.Version.ActiveVersion(ctx)
-	if err := tableDesc.AllocateIDs(ctx, version); err != nil {
-		return err
+	{
+		version := p.ExecCfg().Settings.Version.ActiveVersion(ctx)
+		// Temporarily empty the mutation jobs slice for AllocateIDs to not fail.
+		mutationJobs := tableDesc.MutationJobs
+		tableDesc.MutationJobs = nil
+		if err := tableDesc.AllocateIDs(ctx, version); err != nil {
+			return err
+		}
+		tableDesc.MutationJobs = mutationJobs
```
What's the reason for this change in `sql/truncate.go`? Why isn't it previously a problem?

postamar replied:
Good question. AllocateIDs, which is a bad function but which nobody wants to touch, validates the descriptor. Prior to this call, the Mutations slice was zeroed, but the MutationJobs slice only gets zeroed later on after doing things to the jobs. 

So, this is a bit dirty, but hey, reimplementing TRUNCATE in the declarative schema changer should be easy, so I don't feel bad about it.

postamar commented on pkg/sql/catalog/tabledesc/table_desc.go:
```diff
@@ -74,7 +74,7 @@ func (desc *wrapper) GetPostDeserializationChanges() catalog.PostDeserialization
 func (desc *wrapper) HasConcurrentSchemaChanges() bool {
 	return (desc.DeclarativeSchemaChangerState != nil &&
 		desc.DeclarativeSchemaChangerState.JobID != catpb.InvalidJobID) ||
-		(len(desc.Mutations) > 0 && len(desc.MutationJobs) > 0)
+		len(desc.MutationJobs) > 0
```
I actually forgot about types, but other descriptor kinds don't have mutations, so that's why tables (and types!) are special.

FYI there is no such thing as a "mutation job", just legacy schema changer jobs associated with mutations.

postamar commented:
Ready for another look

Xiang-Gu reviewed (approved):
LGTM

postamar commented:
Thanks for the review!

bors r+
